### DISCUSSION TITLE
VSCODE+ARDUINO-extension (mostly now)

### DISCUSSION THREAD

**Discussion by apws** [SpenceKonde/DxCore#231] [Q&A]
I tried this board (no bootloader) but flashing crashed at some unparsed things 
C:\Users\petra\AppData\Local\Arduino15\packages\megaTinyCore\tools\python3\3.7.2-post1/python3 -u C:\Users\petra\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.4.5/tools/prog.py -t uart -u COM17 -b 230400 -d avr128db48{upload.workaround} --fuses 2:{bootloader.OSCCFG} 6:0b00001100 8:{bootloader.BOOTEND} -fC:\Users\petra\AppData\Local\Temp\arduino_build_310431/CNANO_AVR128DB48_Blink.ino.hex -a write 
Error: cannot parse fuse, 'invalid literal for int() with base 0: '{bootloader.OSCCFG}''
Error: cannot parse fuse, 'invalid literal for int() with base 0: '{bootloader.OSCCFG}''

**Comment by apws** [comment]
to be honest, after selection Curiosity Nano nEDBG programmer and uploading with it, it works - is this expected??

EDIT - ya, its noted in readme, sorry

**Reply by SpenceKonde** [reply]
{bootloader.OSCCFG} means you had a megaTinyCore programmer selected, switched to DxCore and didn't select a valid programme (you'll also notice that tools -> programmer menu is blank when this is happening. That's because the selected programmer is one that isn't shown because it 's not available for the parts you are using.

**Reply by apws** [reply]
ya, thanks, I now understood even how to configure well manually the Curiosity Nano board, found the correct USB-CDC UART pins and this one has safe reset pin on PF6, so no problem with autoreset (sic) with optiboot - BUT problem is that I am testing now deeply combo vscode+arduino and while its PERFECT, far better that arduino ide, just because of stored board preferences in the project folder (whatever.ino + .vscode/arduino.json) - just any board settings are persistent to folder and not needed to manually change that in  every aruduino ide instance/sketch .... but vscode/arduino-ext has still quirks which I am now collecting to report issues - one of them is lack of burning bootloader (okay, for this thing is okay to ONCE use arduino ide and flash it), also serial monitor here behaves differently, it cant catch print() output (flush desn't help too), received bytes are flushed to terminal after first println() only, thats theirs bug... another quirk I see as important is lack of any progress indicator for verifying/compiling to support 8sec optiboot delay on any CNANO (both finally dont reset !!!)

another quirk (probably) in DxCore only is that when I configure board in vscode/arduino-ext, there IS NOT FULL LIST of programmers to select from IDE, only SerialUPDI are listed, so its needed to place "programmer": "DxCore:nedbg" into .vsode/arduino.json, ONCE for folder, where then it all persist, so no big deal.

using vscode/arduino-ext, I am now finally able to quickly test more different arduino boards, as board setting is persistent along with ino in his folder, thats HUGE help ....

fact is, I am not working on any specific complex/usefull project, only testing few boards, including tiny Infineon XMC2Go (XMC1100 Cortex-M0+) 

WHAT really seems as DxCore bug though (after testing the same sketch with 4 different boards - cnano avr 128db48, cnano tiny1627, XMC2Go and Leonardo) is that in DxCore, it seems that Serial does nothing - I have identified CDC pins also (here PB1,PB0) and flashed correct optiboot, but Serial monitor shows nothing in arduino ide, nor in vscode/arduino-ext. Initially I thought that something is very bad  in vscode/arduino, so then I identified that my print(".") are not shown as NEVER flushed without any println() ... but thats really problem of theirs serial monitor.

**Reply by apws** [reply]
Correction: "has safe reset pin on PF6, so no problem with autoreset with optiboot " is nonsense - it seems that the onboardd endbg programmer doesnt reset target before uploading to optiboot, I dont know if this can be somehow forced in avrdude, but finally found that for both Curiosity Nano I have its needed to flash bootloader with 8sec delay and then in vscode/arduino-ext reinsert board/usb during "Veryfiing..." (at least this is much less painful than to catch the right moment in arduino ide...) ... isnt really possible to configure avrdude to WAIT to insert board?? I know there is not serial port visible... but it would be far easier to wait for retries and insert then...

**Reply by apws** [reply]
so... final experience with curiosity nano, no more such boards for arduino (honestly I will be testing them in MPLAB-X soon, so there is the reason I bought them) but its far better to have our simple SerialUPDI/UART programmer with autoreset circuit on custom board, howgh :-)

**Comment by apws** [comment]
another test - (now in vscode+arduino/ms) - I tried to select nEDBG programmer for CNANO AVR128DB48 but there are offered only SerialUPDI options, although in Arduino IDE its possible - for example with CNANO TINY1627 when I DONT select it as "Official Microchip Board" in vscode/arduino but manually, as above then there are ALL programmers listed, including nEDGB ... not sure where is the problem

EDIT: placing "programmer": "DxCore:nedbg" into .vscode/arduino.json helps 
(may be its fault of the extension, now as I am again trying avr128db48, NO PROGRAMMER can be selected by menu, but this change to arduino.json works)


VSCODE+ARDUINO
==================
WELL, I understand that you dont know vscode yet, but I highly encourage you at least to TRY IT, quickly. This way with plain "ms arduino" extension its far easier than platformio (not much complex too) and can serve as introduction to vscode - I also dont/didnt like vscode much, as its nodejs based (new arudino ide20 also, though) and feels too "raw" but here this simple combo "clean vscode+arduino extension" may serve as start, even to platformio - and for me it was that... lets try to watch those short videos, this is deadly easy thing for few minutes of fresh relax only :-)

(honestly I have immediately changed File/Preferences/ColorTheme to "Light+ (default)" as I dont like that nighly modes with reversed background context... )
After installing arduino extension (see videos), most important is F1/"Ctrl+Shift+P" command pallete and following entry of things with keyboard (searching board names, libraries, support packages.... everything). - after searching command (try start with "arduino" or "board" etc...), command often shows another panel identical to Arduino IDE (it in fact fetches whats already installed there in Arduino IDE) There are also buttons on bottom bar and build/upload button on right top, but really everything can be done almost immediatelly from keyboard after F1/Ctrl+Shift+P ...

Probably MOST important thing is that folder initialized with .vscode/arduino.json persist settings for ino, so makes "project" - its immediately possible to upload with defined params, good for testing many different boards with example code too.

But FIRST thing is to to F1/Ctrl+Shift+P "settings" (Preferences: Open Settings (UI)) then inside settings search for "arduino", then scroll down to arduino path and enter here path to installation "c:\Program Files (x86)\Arduino" (only FIRST czech video mentions it !!!)

Some videos are using linux, some are mostly about buttons, etc... but combination and trying helps to manage that. And I can say, I feel quite better in that vscode env now .........

start here with this: (its in czech, but WATCH whats happening, the install extension and SETTING PATH, !important)
https://www.youtube.com/watch?v=LI5L2Rf2_fs

this one is also quite short and simple start too (here during vscode setup, !usefull to add vscode to shell context menu !!!)
https://www.youtube.com/watch?v=WVZxK2MEbE4

and just another 3 minutes repeat
https://www.youtube.com/watch?v=VfLTZcKCGfk

this video is with shitty music, but also good tips ("output": "build" in my case to persist output in folder)
(he doesnt check adding vscode to shell context menu - this is GREAT thing - you also dont need to name INO with same name as folder!)
https://www.youtube.com/watch?v=HPUTKj9ddKA

... I originally thought to add this as separate discussion, feel free to do so ... 
vscode is really nice for arduino, and this can serve as start to platformio, which is even better thing to know ... I will follow there too

cheers,
Petr

**Reply by apws** [reply]
ya, now its already clear that using vscode/arduino-ext saves time to test things of regular arduino ide setup - new 2.0rc doesnt help, most important thing are persistent settings in vscode... howgh

**Comment by apws** [comment]
And one thing related to this ... I tried following sketch on 3 boards (Leonardo, XMC2Go (Infineon XMC1100 Cortex-M0+) and microchip Curiosity Nano Tiny1627 ... after some weirdness about LONG delay required after Serial.begin(), I managed to work this in both Arduino IDE and vscode+arduino (dont know if there wasnt some busy com port issues between both ides etc) but  CNANO 1627 (megaTinyCore) reports that "Serial is undefined" to intellisense (and its question if its on your or theirs side....)

Code builds and uploads and works, but there is reported problem - weird that on different board packages it doesnt happen...

![image](https://user-images.githubusercontent.com/10888697/149710924-3dea7e5a-e666-4dba-b13a-18a4915201fa.png)
[main.txt](https://github.com/SpenceKonde/DxCore/files/7879115/main.txt)

**Reply by apws** [reply]
I reported this to ms vscode repo, they workarounded some workarounds recently and definitelly there is something bad, but may be that working with vscode/arduino-ext may help YOU to identify some bugs faster too ...

**Comment by SpenceKonde** [comment]
VScode seems to have a lot of issues with detecting things that are and aren;'t defined and showing warnings like that for stuff that compiles fine. This isn''t the first. 

I'm sure the problem is that Serial is #defined  as Serial0 (that could be overridden by variant or boards.txt entry) and it;'s not properly handling the preprocessor directives in it's warnings.

**Reply by apws** [reply]
okay, ya, this is weird but ball is on side of arudino-ext guys. vscode itself is not culprit but behavior of extensions, and thats what I complain about all the things based on vscode, in case there is too much extensions installed, may be conflicting - this is also what I till now investigated about platformio, which I dont like too much too, yet - ya, I want to test more different chips beyond arduino so wish to manage it, but also use this platformio beast rather through CLion IDE, which is nice and clean real desktop IDE, contrary to vscode, which I understand more as advanced notepad... BUT, in case you have clean instance of vscode and ONLY arduino-extension, this combo works very well and BINDS to existing installation of regular arduino ide and tooling, it takes whats installed and wraps it nicely and I like the persistence of board settings (lots of them often difficult to set correctly in arduino ide while switching chips...) together with ino sketch, thats what I like about this.

Things about Serial are weird and I will try to report the bugs to them, the same as need to repair "serial monitor" which for example cant catch print() by characters without newline....

**Comment by SpenceKonde** [comment]
I cannot fathom why people seem so attached to the serial monitors that come bundled with IDEs. They all suck. 

Use a real full featured serial terminal, and you'll never want to use an IDE's serial monitor again. I use hTerm.....

**Reply by apws** [reply]
thanks, ya

**Reply by apws** [reply]
thanks, I tried hterm, but didnt found way to use here some interactive terminal feature, so tried also "termite", which more resembles arduino ide, having command line on bottom (better) and also remembering enetered strings, quite nice, for those interested
https://www.compuphase.com/software_termite.htm

**Comment by SpenceKonde** [comment]
Hm, how does one get Termite to display text in hex and ascii (both at the same time)? That for me is the killer feature of hterm.
